Error in stage [HttpRequestParser]

Hi, I am using akka-http/10.0.9 and on production I see following exceptions (not very often, probably every 500k requests):

```
Error in stage [HttpRequestParser]: 762 
java.lang.ArrayIndexOutOfBoundsException: 762
        at akka.http.impl.engine.parsing.HttpHeaderParser.parseHeaderValue(HttpHeaderParser.scala:194)
        at akka.http.impl.engine.parsing.HttpHeaderParser.parseHeaderLine(HttpHeaderParser.scala:130)
        at akka.http.impl.engine.parsing.HttpMessageParser$class.parseHeaderLines(HttpMessageParser.scala:125)
        at akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.parseHeaderLines(HttpRequestParser.scala:44)
        at akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.parseMessage(HttpRequestParser.scala:77)
        at akka.http.impl.engine.parsing.HttpMessageParser$class.startNewMessage(HttpMessageParser.scala:100)
        at akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.startNewMessage(HttpRequestParser.scala:44)
        at akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$startNewMessage$1.apply(HttpMessageParser.scala:101)
        at akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$startNewMessage$1.apply(HttpMessageParser.scala:101)
        at akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$continue$2.apply(HttpMessageParser.scala:264)
        at akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$continue$2.apply(HttpMessageParser.scala:264)
        at akka.http.impl.engine.parsing.HttpMessageParser$class.run$1(HttpMessageParser.scala:64)
        at akka.http.impl.engine.parsing.HttpMessageParser$class.parseBytes(HttpMessageParser.scala:78)
        at akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.parseBytes(HttpRequestParser.scala:44)
        at akka.http.impl.engine.parsing.HttpMessageParser$class.parseSessionBytes(HttpMessageParser.scala:60)
        at akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.parseSessionBytes(HttpRequestParser.scala:44)
        at akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.onPush(HttpRequestParser.scala:55)
        at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:747)
        at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:649)
        at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:471)
        at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:423)
        at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:603)
        at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:618)
        at akka.actor.Actor$class.aroundReceive(Actor.scala:502)
        at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:529)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
        at akka.actor.ActorCell.invoke_aroundBody0(ActorCell.scala:495)
        at akka.actor.ActorCell$AjcClosure1.run(ActorCell.scala:1)
        at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149)
        at akka.kamon.instrumentation.ActorMonitors$$anon$1$$anonfun$processMessage$1.apply(ActorMonitor.scala:63)
        at kamon.trace.Tracer$.withContext(TracerModule.scala:58)
        at akka.kamon.instrumentation.ActorMonitors$$anon$1.processMessage(ActorMonitor.scala:62)
        at akka.kamon.instrumentation.ActorCellInstrumentation.aroundBehaviourInvoke(ActorInstrumentation.scala:45)
        at akka.actor.ActorCell.invoke(ActorCell.scala:488)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
        at akka.dispatch.Mailbox.run(Mailbox.scala:224)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

```

and sometimes:

```
Error in stage [HttpRequestParser]: null 
java.nio.BufferOverflowException: null
        at java.nio.Buffer.nextPutIndex(Buffer.java:521)
        at java.nio.HeapByteBuffer.put(HeapByteBuffer.java:169)
        at akka.http.impl.engine.parsing.HttpHeaderParser$.akka$http$impl$engine$parsing$HttpHeaderParser$$scanHeaderValue(HttpHeaderParser.scala:557)
        at akka.http.impl.engine.parsing.HttpHeaderParser$ModeledHeaderValueParser.apply(HttpHeaderParser.scala:499)
        at akka.http.impl.engine.parsing.HttpHeaderParser.parseAndInsertHeader$1(HttpHeaderParser.scala:173)
        at akka.http.impl.engine.parsing.HttpHeaderParser.parseHeaderValue(HttpHeaderParser.scala:187)
        at akka.http.impl.engine.parsing.HttpHeaderParser.parseHeaderLine(HttpHeaderParser.scala:130)
        at akka.http.impl.engine.parsing.HttpMessageParser$class.parseHeaderLines(HttpMessageParser.scala:125)
        at akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.parseHeaderLines(HttpRequestParser.scala:44)
        at akka.http.impl.engine.parsing.HttpMessageParser$class.akka$http$impl$engine$parsing$HttpMessageParser$$parseHeaderLinesAux(HttpMessageParser.scala:170)
        at akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$parseHeaderLines$1.apply(HttpMessageParser.scala:131)
        at akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$parseHeaderLines$1.apply(HttpMessageParser.scala:131)
        at akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$continue$1.apply(HttpMessageParser.scala:263)
        at akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$continue$1.apply(HttpMessageParser.scala:263)
        at akka.http.impl.engine.parsing.HttpMessageParser$class.run$1(HttpMessageParser.scala:64)
        at akka.http.impl.engine.parsing.HttpMessageParser$class.parseBytes(HttpMessageParser.scala:78)
        at akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.parseBytes(HttpRequestParser.scala:44)
        at akka.http.impl.engine.parsing.HttpMessageParser$class.parseSessionBytes(HttpMessageParser.scala:60)
        at akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.parseSessionBytes(HttpRequestParser.scala:44)
        at akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.onPush(HttpRequestParser.scala:55)
        at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:747)
        at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:649)
        at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:471)
        at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:423)
        at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:603)
        at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:618)
        at akka.actor.Actor$class.aroundReceive(Actor.scala:502)
        at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:529)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
        at akka.actor.ActorCell.invoke_aroundBody0(ActorCell.scala:495)
        at akka.actor.ActorCell$AjcClosure1.run(ActorCell.scala:1)
        at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149)
        at akka.kamon.instrumentation.ActorMonitors$$anon$1$$anonfun$processMessage$1.apply(ActorMonitor.scala:63)
        at kamon.trace.Tracer$.withContext(TracerModule.scala:58)
        at akka.kamon.instrumentation.ActorMonitors$$anon$1.processMessage(ActorMonitor.scala:62)
        at akka.kamon.instrumentation.ActorCellInstrumentation.aroundBehaviourInvoke(ActorInstrumentation.scala:45)
        at akka.actor.ActorCell.invoke(ActorCell.scala:488)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
        at akka.dispatch.Mailbox.run(Mailbox.scala:224)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```
and then lot of these:

```
Error in stage [HttpRequestParser]: null 
java.lang.ArrayIndexOutOfBoundsException: null
(no stack trace)
```


I was not able to find the offending request (I have tcpdump but nothing special out there, nginx which run in front does not report errors neither).

BTW, I have seen https://github.com/akka/akka/issues/18397 and I will try to disable cache and see if it helps.

Here is another ArrayIndexOutOfBoundsException:

<details>

```
path("upload") {
  post {
    fileUploadAll("test") { uploads =>
      complete(StatusCodes.OK)
    }
  }
}
```

```
% echo $LANG
ru_RU.UTF-8
% touch А
% touch Б
% curl 'http://127.0.0.1:8080/upload' -F "test=@А" -F "test=@Б" --trace-ascii -
== Info:   Trying 127.0.0.1:8080...
== Info: TCP_NODELAY set
== Info: Connected to 127.0.0.1 (127.0.0.1) port 8080 (#0)
=> Send header, 192 bytes (0xc0)
0000: POST /upload HTTP/1.1
0017: Host: 127.0.0.1:8080
002d: User-Agent: curl/7.67.0
0046: Accept: */*
0053: Content-Length: 342
0068: Content-Type: multipart/form-data; boundary=--------------------
00a8: ----8db5165dfc64f3df
00be: 
=> Send data, 342 bytes (0x156)
0000: --------------------------8db5165dfc64f3df
002c: Content-Disposition: form-data; name="test"; filename=".."
0068: Content-Type: application/octet-stream
0090: 
0092: 
0094: --------------------------8db5165dfc64f3df
00c0: Content-Disposition: form-data; name="test"; filename=".."
00fc: Content-Type: application/octet-stream
0124: 
0126: 
0128: --------------------------8db5165dfc64f3df--
== Info: We are completely uploaded and fine
== Info: Mark bundle as not supporting multiuse
<= Recv header, 26 bytes (0x1a)
0000: HTTP/1.1 400 Bad Request
<= Recv header, 27 bytes (0x1b)
0000: Server: akka-http/10.1.11
<= Recv header, 37 bytes (0x25)
0000: Date: Tue, 11 Feb 2020 22:50:51 GMT
<= Recv header, 41 bytes (0x29)
0000: Content-Type: text/plain; charset=UTF-8
<= Recv header, 20 bytes (0x14)
0000: Content-Length: 73
<= Recv header, 2 bytes (0x2)
0000: 
<= Recv data, 73 bytes (0x49)
0000: The request content was malformed:.Index 764 out of bounds for l
0040: ength 762
The request content was malformed:
Index 764 out of bounds for length 762== Info: Connection #0 to host 127.0.0.1 left intact
```

```
[2020-02-12 01:50:51,912] [ERROR] [akka.actor.RepointableActorRef] [HelloAkkaHttpServer-akka.actor.default-dispatcher-23] [akka://HelloAkkaHttpServer/system/Materializers/StreamSupervisor-0/flow-20-0-headSink] - Error in stage [akka.http.impl.engine.parsing.BodyPartParser@381a7557]: Index 764 out of bounds for length 762
java.lang.ArrayIndexOutOfBoundsException: Index 764 out of bounds for length 762
	at akka.http.impl.engine.parsing.HttpHeaderParser.parseHeaderValue(HttpHeaderParser.scala:195)
	at akka.http.impl.engine.parsing.HttpHeaderParser.parseHeaderLine(HttpHeaderParser.scala:131)
	at akka.http.impl.engine.parsing.BodyPartParser$$anon$1.parseHeaderLines(BodyPartParser.scala:155)
	at akka.http.impl.engine.parsing.BodyPartParser$$anon$1.$anonfun$parseEntity$1(BodyPartParser.scala:215)
	at akka.http.impl.engine.parsing.BodyPartParser$$anon$1.loop$1(BodyPartParser.scala:98)
	at akka.http.impl.engine.parsing.BodyPartParser$$anon$1.run(BodyPartParser.scala:103)
	at akka.http.impl.engine.parsing.BodyPartParser$$anon$1.onPush(BodyPartParser.scala:66)
	at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:541)
	at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:527)
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:390)
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:624)
	at akka.stream.impl.fusing.GraphInterpreterShell.init(ActorGraphInterpreter.scala:594)
	at akka.stream.impl.fusing.ActorGraphInterpreter.tryInit(ActorGraphInterpreter.scala:701)
	at akka.stream.impl.fusing.ActorGraphInterpreter.preStart(ActorGraphInterpreter.scala:750)
	at akka.actor.Actor.aroundPreStart(Actor.scala:545)
	at akka.actor.Actor.aroundPreStart$(Actor.scala:545)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundPreStart(ActorGraphInterpreter.scala:690)
	at akka.actor.ActorCell.create(ActorCell.scala:637)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:509)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:531)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:294)
	at akka.dispatch.Mailbox.run(Mailbox.scala:229)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:242)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```
</details>

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpHeaderParser.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpHeaderParser.scala
@@ -211,8 +211,8 @@ private[engine] final class HttpHeaderParser private (
   @tailrec
   private def insert(input: ByteString, value: AnyRef)(cursor: Int = 0, endIx: Int = input.length, nodeIx: Int = 0, colonIx: Int = 0): Unit = {
     val char =
-      if (cursor < colonIx) CharUtils.toLowerCase(input(cursor).toChar)
-      else if (cursor < endIx) input(cursor).toChar
+      if (cursor < colonIx) CharUtils.toLowerCase((input(cursor) & 0xff).toChar)
+      else if (cursor < endIx) (input(cursor) & 0xff).toChar
       else '\u0000'
     val node = nodes(nodeIx)
     if (char == node) insert(input, value)(cursor + 1, endIx, nodeIx + 1, colonIx) // fast match, descend into only subnode
@@ -258,7 +258,7 @@ private[engine] final class HttpHeaderParser private (
   private def insertRemainingCharsAsNewNodes(input: ByteString, value: AnyRef)(cursor: Int = 0, endIx: Int = input.length, valueIx: Int = newValueIndex, colonIx: Int = 0): Unit = {
     val newNodeIx = newNodeIndex
     if (cursor < endIx) {
-      val c = input(cursor).toChar
+      val c = (input(cursor) & 0xff).toChar
       val char = if (cursor < colonIx) CharUtils.toLowerCase(c) else c
       nodes(newNodeIx) = char
       insertRemainingCharsAsNewNodes(input, value)(cursor + 1, endIx, valueIx, colonIx)

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/parsing/package.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/parsing/package.scala
@@ -26,7 +26,7 @@ package object parsing {
     case x                              => x.toString
   }
 
-  private[http] def byteChar(input: ByteString, ix: Int): Char = byteAt(input, ix).toChar
+  private[http] def byteChar(input: ByteString, ix: Int): Char = (byteAt(input, ix) & 0xff).toChar
 
   private[http] def byteAt(input: ByteString, ix: Int): Byte =
     if (ix < input.length) input(ix) else throw NotEnoughDataException

--- a/akka-http-core/src/test/scala/akka/http/impl/engine/parsing/HttpHeaderParserSpec.scala
+++ b/akka-http-core/src/test/scala/akka/http/impl/engine/parsing/HttpHeaderParserSpec.scala
@@ -181,6 +181,16 @@ abstract class HttpHeaderParserSpec(mode: String, newLine: String) extends WordS
         RawHeader("4-UTF8-Bytes", "Surrogate pairs: \uD801\uDC1B\uD801\uDC04\uD801\uDC1B!")
     }
 
+    "parse multiple header lines subsequently with UTF-8 characters one after another without crashing" in new TestSetup {
+      parseLine(s"""Content-Disposition: form-data; name="test"; filename="λ"${newLine}x""")
+      // The failing parsing line is one that must share a prefix with the utf-8 line up to the non-ascii char. The next character
+      // doesn't even have to be a non-ascii char.
+      parseLine(s"""Content-Disposition: form-data; name="test"; filename="test"${newLine}x""")
+      // But it could be
+      parseLine(s"""Content-Disposition: form-data; name="test"; filename="Б"${newLine}x""")
+
+    }
+
     "produce an error message for lines with an illegal header name" in new TestSetup() {
       the[ParsingException] thrownBy parseLine(s" Connection: close${newLine}x") should have message "Illegal character ' ' in header name"
       the[ParsingException] thrownBy parseLine(s"Connection : close${newLine}x") should have message "Illegal character ' ' in header name"

